_posts/2020-08-24-training-imagenet-with-r/training-imagenet-with-r.Rmd

```diff
@@ -0,0 +1,270 @@
+
+Once we've chosen a model, we will want to me make sure that it properly trains on a subset of ImageNet:
+
+```{r}
+remotes::install_github("r-tensorflow/alexnet")
+alexnet::alexnet_train(data = data)
+```
+```
+Epoch 1/2
+ 103/2269 [>...............] - ETA: 5:52 - loss: 72306.4531 - accuracy: 0.9748
+```
+
+So far so good! However, this post is about enabling large-scale training across multiple GPUs, so we want to make sure we are using as many as we can. Unfortunately, running `nvidia-smi` will show that only one GPU currently being used:
+
+```{bash eval=FALSE}
+nvidia-smi
+```
+```
++-----------------------------------------------------------------------------+
+| NVIDIA-SMI 418.152.00   Driver Version: 418.152.00   CUDA Version: 10.1     |
+|-------------------------------+----------------------+----------------------+
+| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
+| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
+|===============================+======================+======================|
+|   0  Tesla K80           Off  | 00000000:00:05.0 Off |                    0 |
+| N/A   48C    P0    89W / 149W |  10935MiB / 11441MiB |     28%      Default |
++-------------------------------+----------------------+----------------------+
+|   1  Tesla K80           Off  | 00000000:00:06.0 Off |                    0 |
+| N/A   74C    P0    74W / 149W |     71MiB / 11441MiB |      0%      Default |
++-------------------------------+----------------------+----------------------+
+                                                                               
++-----------------------------------------------------------------------------+
+| Processes:                                                       GPU Memory |
+|  GPU       PID   Type   Process name                             Usage      |
+|=============================================================================|
++-----------------------------------------------------------------------------+
+```
+
+In order to train across multiple GPUs, we need to define a distributed-processing strategy. If this is a new concept, it might be a good time to take a look at the [Distributed Training with Keras](https://tensorflow.rstudio.com/tutorials/advanced/distributed/distributed_training_with_keras/) tutorial and the [distributed training with TensorFlow](https://www.tensorflow.org/guide/distributed_training) docs. Or, if you allow us to overly-simplify the process, all you have to do is define and compile your model under the right scope. A step-by-step explanation is available in the [Distributed Deep Learning with TensorFlow and R](https://www.youtube.com/watch?v=DQyLTlD1IBc) video. In this case, the `alexnet` model [already supports](https://github.com/r-tensorflow/alexnet/blob/57546/R/alexnet_train.R#L92-L94) a strategy parameter, so all we have to do is pass it along.
```

Comment:
```suggestion
In order to train across multiple GPUs, we need to define a distributed-processing strategy. If this is a new concept, it might be a good time to take a look at the [Distributed Training with Keras](https://tensorflow.rstudio.com/tutorials/advanced/distributed/distributed_training_with_keras/) tutorial and the [distributed training with TensorFlow](https://www.tensorflow.org/guide/distributed_training) docs. Or, if you allow us to oversimplify the process, all you have to do is define and compile your model under the right scope. A step-by-step explanation is available in the [Distributed Deep Learning with TensorFlow and R](https://www.youtube.com/watch?v=DQyLTlD1IBc) video. In this case, the `alexnet` model [already supports](https://github.com/r-tensorflow/alexnet/blob/57546/R/alexnet_train.R#L92-L94) a strategy parameter, so all we have to do is pass it along.
```